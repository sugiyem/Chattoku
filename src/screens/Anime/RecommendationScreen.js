import React, { useEffect, useState } from "react";
import { Alert, StyleSheet, TouchableOpacity, View } from "react-native";
import { Button, ButtonText, Container } from "../../styles/GeneralStyles";
import FetchUserInfo from "../../services/Profile/FetchUserInfo";
import Loading from "../../components/Miscellaneous/Loading";

const RecommendationScreen = ({ navigation }) => {
  const [genres, setGenres] = useState([]);
  const [isGenreLoaded, setIsGenreLoaded] = useState(false);

  useEffect(() => {
    setIsGenreLoaded(false);

    return FetchUserInfo({
      onSuccesfulFetch: (data) => {
        setGenres(data.genres);
        setIsGenreLoaded(true);
      },
      onFailure: (error) => {
        Alert.alert("Error", error.message);
      }
    });
  }, []);

  const isPageLoading = !isGenreLoaded;

  const RenderPage = () => (
    <Container>
      <TouchableOpacity
        style={styles.backButton}
        onPress={() => navigation.goBack()}
      >
        <ButtonText>Go Back</ButtonText>
      </TouchableOpacity>

      <View style={styles.buttonsContainer}>
        <Button onPress={() => navigation.navigate("FavoriteAnime")} pad="20px">
          <ButtonText size="20px">
            Check recommendation from your favorite anime
          </ButtonText>
        </Button>

        <Button
          onPress={() =>
            navigation.navigate("FavGenreRecommendation", {
              genres: genres
            })
          }
          pad="20px"
        >
          <ButtonText size="20px">
            Check recommendation from your favorite genre
          </ButtonText>
        </Button>
      </View>
    </Container>
  );

  return <Loading isLoading={isPageLoading} ExpectedRender={RenderPage} />;
};

export default RecommendationScreen;

const styles = StyleSheet.create({
  backButton: {
    alignSelf: "stretch",
    justifyContent: "flex-start",
    borderRadius: 10,
    borderWidth: 1,
    backgroundColor: "cyan",
    margin: 5,
    padding: 5
  },
  buttonsContainer: {
    justifyContent: "center",
    flex: 1
  }
});
