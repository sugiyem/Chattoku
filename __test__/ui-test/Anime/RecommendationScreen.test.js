import React, { useEffect } from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react-native";
import RecommendationScreen from "../../../src/screens/Anime/RecommendationScreen";

const mockNavigate = jest.fn();
const mockSetState = jest.fn();
const mockBackNavigate = jest.fn();
const mockNavigation = {
  goBack: mockBackNavigate,
  navigate: mockNavigate
};
const mockData = ["Action", "Fantasy", "Isekai"];

jest.mock("@react-navigation/core", () => ({
  useNavigation: () => mockNavigation
}));

jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: (initial) => [
    typeof initial === "boolean" ? true : mockData,
    mockSetState
  ],
  useEffect: () => {}
}));

describe("Test RecommendationScreen UI", () => {
  it("Renders Correctly", () => {
    const tree = renderer
      .create(<RecommendationScreen navigation={mockNavigation} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Navigation Works", () => {
    const { getByTestId } = render(
      <RecommendationScreen navigation={mockNavigation} />
    );

    fireEvent.press(getByTestId("backButton"));
    expect(mockBackNavigate).toHaveBeenCalled();

    fireEvent.press(getByTestId("genreButton"));
    expect(mockNavigate).toHaveBeenCalledWith("FavGenreRecommendation", {
      genres: mockData
    });

    fireEvent.press(getByTestId("animeButton"));
    expect(mockNavigate).toHaveBeenCalledWith("FavoriteAnime");
  });
});
