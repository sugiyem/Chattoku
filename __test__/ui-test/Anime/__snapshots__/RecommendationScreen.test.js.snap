// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RecommendationScreen UI Renders Correctly 1`] = `
<ViewManagerAdapter_ExpoLinearGradient
  proxiedProperties={
    Object {
      "colors": Array [
        4287090411,
        4283256319,
      ],
      "endPoint": Array [
        1,
        1,
      ],
      "locations": undefined,
      "startPoint": Array [
        0,
        0,
      ],
    }
  }
  style={
    Object {
      "flex": 1,
      "padding": 5,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "cyan",
        "borderRadius": 10,
        "borderWidth": 1,
        "justifyContent": "flex-start",
        "marginBottom": 5,
        "marginLeft": 5,
        "marginRight": 5,
        "marginTop": 5,
        "opacity": 1,
        "paddingBottom": 5,
        "paddingLeft": 5,
        "paddingRight": 5,
        "paddingTop": 5,
      }
    }
    testID="backButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#483d8b",
            "fontSize": 14,
            "textAlign": "center",
          },
        ]
      }
    >
      Go Back
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#44d0fe",
          "borderColor": "navy",
          "borderRadius": 20,
          "borderWidth": 2,
          "marginBottom": 15,
          "marginLeft": 15,
          "marginRight": 15,
          "marginTop": 15,
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
        }
      }
      testID="animeButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "navy",
              "fontSize": 22,
              "textAlign": "center",
            },
          ]
        }
      >
        Check recommendation from your favorite anime
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#44d0fe",
          "borderColor": "navy",
          "borderRadius": 20,
          "borderWidth": 2,
          "marginBottom": 15,
          "marginLeft": 15,
          "marginRight": 15,
          "marginTop": 15,
          "opacity": 1,
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
        }
      }
      testID="genreButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "navy",
              "fontSize": 22,
              "textAlign": "center",
            },
          ]
        }
      >
        Check recommendation from your favorite genre
      </Text>
    </View>
  </View>
</ViewManagerAdapter_ExpoLinearGradient>
`;
